sinks/metrics/metrics.go

```diff
@@ -0,0 +1,106 @@
+// Package metrics provides sinks that are used by veneur internally.
+package metrics
+
+import (
+	"github.com/sirupsen/logrus"
+	"github.com/stripe/veneur/protocol"
+	"github.com/stripe/veneur/samplers"
+	"github.com/stripe/veneur/sinks"
+	"github.com/stripe/veneur/ssf"
+	"github.com/stripe/veneur/trace"
+)
+
+type metricExtractionSink struct {
+	workers                []Processor
+	indicatorSpanTimerName string
+	log                    *logrus.Logger
+}
+
+var _ sinks.SpanSink = &metricExtractionSink{}
+
+// Processor represents a thing that can process UDPMetrics.
+type Processor interface {
+	// IngestUDP takes a single UDPMetric and processes it in the worker.
+	IngestUDP(samplers.UDPMetric)
+}
+
+// NewMetricExtractionSink sets up and creates a span sink that
+// extracts metrics ("samples") from SSF spans and reports them to a
+// veneur's metrics workers.
+func NewMetricExtractionSink(mw []Processor, timerName string, log *logrus.Logger) (sinks.SpanSink, error) {
+	return &metricExtractionSink{
+		workers:                mw,
+		indicatorSpanTimerName: timerName,
+		log: log,
+	}, nil
+}
+
+// Name returns "metric_extraction".
+func (m metricExtractionSink) Name() string {
+	return "metric_extraction"
+}
+
+// Start is a no-op.
+func (m metricExtractionSink) Start(*trace.Client) error {
+	return nil
+}
+
+func (m *metricExtractionSink) sendMetrics(metrics []samplers.UDPMetric) {
+	for _, metric := range metrics {
+		m.workers[metric.Digest%uint32(len(m.workers))].IngestUDP(metric)
+	}
+}
+
+func (m *metricExtractionSink) sendSample(sample *ssf.SSFSample) error {
+	metric, err := samplers.ParseMetricSSF(sample)
+	if err != nil {
+		return err
+	}
+	m.sendMetrics([]samplers.UDPMetric{metric})
+	return nil
+}
+
+// Ingest extracts metrics from an SSF span, and feeds them into the
+// appropriate metric sinks.
+func (m metricExtractionSink) Ingest(span *ssf.SSFSpan) error {
+	metrics, err := samplers.ConvertMetrics(span)
+	if err != nil {
+		if _, ok := err.(samplers.InvalidMetrics); ok {
+			m.log.WithError(err).
+				Warn("Could not parse metrics from SSF Message")
+			m.sendSample(ssf.Count("veneur.ssf.error_total", 1, map[string]string{
+				"packet_type": "ssf_metric",
+				"step":        "extract_metrics",
+				"reason":      "invalid_metrics",
+			}))
+		} else {
+			m.log.WithError(err).Error("Unexpected error extracting metrics from SSF Message")
+			m.sendSample(ssf.Count("veneur.ssf.error_total", 1, map[string]string{
+				"packet_type": "ssf_metric",
+				"step":        "extract_metrics",
+				"reason":      "unexpected_error",
+				"error":       err.Error(),
+			}))
+			return err
+		}
+	}
+	m.sendMetrics(metrics)
+
+	if err := protocol.ValidateTrace(span); err != nil {
+		return err
+	}
+
+	indicatorMetrics, err := samplers.ConvertIndicatorMetrics(span, m.indicatorSpanTimerName)
```

Comment:
I was going to propose this be configurable, but based on our chat today I think we want to embrace the opinions that Veneur represents: span-centric instrumentation. Leaving it configurable allows users to get the wrong impression of where we want to go.